src/types/infer.ts

```diff
@@ -0,0 +1,1359 @@
+            return new TypeNameType(new UserDefinedType(getFQDefName(def), def));
+        }
+
+        throw new Error(
+            `NYI infer of identifier ${node.name} (${pp(node)}) with def ${def.constructor.name}`
+        );
+    }
+
+    typeOfLiteral(node: Literal): TypeNode {
+        if (node.kind === "number") {
+            let val = new Decimal(node.value);
+
+            if (node.subdenomination !== undefined) {
+                val = val.times(subdenominationMultipliers[node.subdenomination]);
+            }
+
+            return new IntLiteralType(val);
+        }
+
+        if (node.kind === "string" || node.kind === "unicodeString" || node.kind === "hexString") {
+            const [val, kind]: [string, StringLiteralKind] =
+                node.kind === "hexString" || node.value === null
+                    ? [node.hexValue, "hexString"]
+                    : [node.value, node.kind];
+
+            return new StringLiteralType(val, kind);
+        }
+
+        return types.bool;
+    }
+
+    typeOfMemberAccess(node: MemberAccess): TypeNode {
+        const baseT = this.typeOf(node.vExpression);
+
+        if (baseT instanceof PointerType) {
+            const toT = baseT.to;
+
+            /// Fields of structs
+            if (toT instanceof UserDefinedType && toT.definition instanceof StructDefinition) {
+                const fields = toT.definition.vMembers.filter(
+                    (fieldDef) => fieldDef.name === node.memberName
+                );
+
+                if (fields.length !== 1) {
+                    throw new SolTypeError(
+                        `No field ${node.memberName} found on struct ${toT.definition.name} in ${pp(
+                            node
+                        )}`
+                    );
+                }
+
+                assert(fields[0].vType !== undefined, ``);
+                return specializeType(typeNameToTypeNode(fields[0].vType), baseT.location);
+            }
+
+            /// Fields on contract vars. Should always be a function
+            if (toT instanceof UserDefinedType && toT.definition instanceof ContractDefinition) {
+                const contract = toT.definition;
+                const res = this.typeOfResolved(node.memberName, contract, true);
+
+                if (res === undefined) {
+                    throw new SolTypeError(
+                        `No field ${node.memberName} found on contract ${contract.name} in ${pp(
+                            node
+                        )}`
+                    );
+                }
+
+                return res;
+            }
+
+            if (toT instanceof ArrayType) {
+                if (node.memberName === "length") {
+                    return types.uint;
+                }
+
+                if (node.memberName === "push") {
+                    return new BuiltinFunctionType(undefined, [toT.elementT], []);
+                }
+            }
```

Comment:
I guess `pop()` is missing here. See https://docs.soliditylang.org/en/v0.8.15/types.html#array-members

Reply:
Fixed in [51bdea0](https://github.com/ConsenSys/solc-typed-ast/pull/142/commits/51bdea0cbff6131127af6ceb6d1e12c11faaecc5).